MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC (#3861)
matrix-org/matrix-doc: proposals/3861-next-generation-auth.md

# MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC

This MSC proposes a new set of authentication APIs for Matrix, based on the OAuth 2.0 and OpenID Connect (OIDC) specifications.

## Motivation

The goal of this MSC is to propose a new set of authentication APIs for Matrix, based on the OAuth 2.0 and OpenID Connect (OIDC) specifications.
To understand this proposal, it is important to lay out the reasoning which led to this conclusion.

This section explains this reasoning before getting into the detailed proposal.
It is purely informative and not part of the actual proposal.
The proposal itself is high-level, and the technical implications are split into separate, more focused MSCs.

### Current paradigm

> The Matrix Client-Server API currently has endpoints for authenticating and managing users.

Reviewer: Done in 9ff541cc

This API design is focused on displaying authentication UI as native controls within the Matrix client.
It is meant to offer multiple server-defined, multi-staged authentication flows, which the client then uses to build a native UI.

As an example, on the Matrix.org homeserver, the user registration flow is gated behind three stages:

1. A CAPTCHA challenge
1. An agreement to the terms of service
1. An email verification

On paper, this lets clients implement their own UI for authentication, but in practice, the dynamic nature of the flows makes it difficult to implement a streamlined UI.

> Moreover, the fact that those stages are displayed on client-driven UI, and not on a homeserver-specific domain, makes them insecure, impossible to design, or behave poorly in terms of user experience.

Reviewer: Is this a hard requirement? I personally prefer using one sentence per line for the reasons outlined here: https://nick.groenen.me/notes/one-sentence-per-line/

> Moreover, the fact that those stages are displayed on client-driven UI, and not on a homeserver-specific domain, makes them insecure, impossible to design, or behave poorly in terms of user experience.

Reviewer 2: not a hard requirement, provided the lines are somewhat short. It's mostly to avoid weirdness with github diffs when small changes bury conversations.

- There is no built-in way to keep state throughout the flow, making it hard for the client to distinguish multiple concurrent flows, and potentially leaking the "login token" to the wrong homeserver.

### OAuth 2.0 and OpenID Connect as building blocks

Quoting the [specification](https://spec.matrix.org/latest/#introduction-to-the-matrix-apis):

> Matrix is a set of open APIs for open-federated Instant Messaging (IM), Voice over IP (VoIP) and Internet of Things (IoT) communication, designed to create and support a new global real-time communication ecosystem. The intention is to provide an **open decentralised pubsub layer for the internet for securely persisting and publishing/subscribing JSON objects**.

Fundamentally, Matrix does not set out to be an authentication protocol.
The ecosystem needs authentication to work, but it is not core to the mission.

This MSC essentially proposes building on top of well-established authentication protocols, defined by OAuth 2.0 RFCs and OpenID Connect specifications.

OAuth 2.0 is a framework for building authorization systems, defined across multiple RFCs by the IETF.

> OpenID Connect is an effort by the OpenID Foundation to standardize on top of OAuth 2.0.

Reviewer: Done in 2acf4387

Most identity providers are also designed to provide service to a fixed set of applications, which does not fit the Matrix ecosystem, where users can use any number of different clients.

This means that backfitting Matrix-specific concepts on top of OpenID Connect would be a bad idea, especially as one important goal of this proposal is to keep the current authentication paradigm working for some time.

**Note**: an earlier version of this MSC focused on 'delegating' authentication to an identity provider, but it showed its limitations and added much confusion over the intent of the proposal.

## Proposal

This proposal introduces a new set of authentication APIs for Matrix, based on OAuth 2.0 and OpenID Connect (OIDC) specifications.

As a first step, it introduces those APIs as altenratives to the existing User-Interactive Authentication (UIA) APIs, acknowledging the complexity of covering all the use cases of the existing APIs.

> The long-term goal is to deprecate the existing UIA APIs and replace them with the new OAuth 2.0/OIDC-based APIs.

Reviewer 3: Since this MSC explicitly doesn't deprecate anything (see below), I suppose the details of the eventual deprecation don't need to be discussed here any more?

> The long-term goal is to deprecate the existing UIA APIs and replace them with the new OAuth 2.0/OIDC-based APIs.

Reviewer: I've reworded that a little bit in f398886b

We plan to deprecate the existing account management APIs and UIA-protected endpoints, replacing them as needed. As we don't yet deprecate, there is no exhaustive list of what we deprecate

> The long-term goal is to deprecate the existing UIA APIs and replace them with the new OAuth 2.0/OIDC-based APIs.

Reviewer 4: I'm still not convinced that we actually want to deprecate all the endpoints that are currently protected by UIA, rather than just removing the requirement that they be protected by UIA, but I guess it's not important.

### Base authentication flow

To cover the most common use case of authenticating an end-user, the following MSCs are necessary:

- [MSC2964: Usage of OAuth 2.0 authorization code grant and refresh token grant][MSC2964] describes how the main authentication flow works
- [MSC2965: OAuth 2.0 Authorization Server Metadata discovery][MSC2965] describes how a client can discover the authentication server metadata of the homeserver
- [MSC2966: Usage of OAuth 2.0 Dynamic Client Registration][MSC2966] describes how a client can register itself with the homeserver to get a client identifier
- [MSC2967: API scopes][MSC2967] defines the first set of access scopes and the basis for future access scopes
- [MSC4254: Usage of RFC7009 Token Revocation for Matrix client logout][MSC4254] describes how a client can end a client session

### Account management

This moves the user-interface for some account management tasks from the client to the homeserver.

> Existing APIs like `/_matrix/client/v3/capabilities` help clients understand which account-management API endpoints are unavailable, but they don't offer alternatives to a homeserver-provided user-interface.

Reviewer: Done in 0a86b697

> Existing APIs like `/_matrix/client/v3/capabilities` help clients understand which account-management API endpoints are unavailable, but they don't offer alternatives to a homeserver-provided user-interface.

Reviewer 4: still no link here.

> To build this bridge between the client user-interface and the homeserver, [MSC4191: Account management deep-linking][MSC4191] proposes a way to deep-link to the account management capabilities of the homeserver.

Reviewer: Clarified in d76b9a29

### Transition and existing client support

To help client transition to the next-generation auth, this proposal is designed to offer backward-compatible APIs through the `m.login.sso` login flow.

> How this is intended to work, and let client offer reasonable user-experience is covered by [MSC3824: OIDC-aware clients][MSC3824].

Reviewer: I've clarified that those are not required for acceptance of this MSC in d76b9a29

## Sample flow

> This section describes a sample flow, taking together the steps described in more details in the dedicated MSCs.

Reviewer: I've clarified that elsewhere in d76b9a29

This adds complexity for homeserver implementers if they want to support both the old and new APIs.

### Reliance on the system browser

The main authentication flow of this proposal is a web-based flow that is intended to run in the context of a system browser.
There are alternative flows which could help support login on devices where a browser is not available, or for automation purposes, but they are not yet part of this proposal.

Additionally, in some contexts, the system browser can be a security risk, as it has a wide attack surface inherent to its complexity.
This can be mitigated to some extent by having homeservers implement authorization flows that don't require JavaScript, but this doesn't completely eliminate the risk.

### Homeserver implementation complexity

In practice, to provide a good user experience, homeservers have to implement web views for the authentication flows, which is complex to implement well.
This means having proper accessibility, translations, and UX.
Those concerns were previously only affecting client implementations, and will now also affect homeserver implementations.

> On the other hand, the previous registration flow was notoriously complex to implement both for clients and homeservers, and this proposal removes a lot of that complexity from the client side.

Reviewer: That's a good point, I've added words about mentions of UIA fallbacks in f48fb925

| [MSC3782: Matrix public key login spec][MSC3782]                                                                                | Similar to above                                                                                                                                                                                                                                                                                                           |
| [MSC3744: Support for flexible authentication][MSC3744]                                                                         | OIDC would instead be used as the pluggable layer for auth in Matrix                                                                                                                                                                                                                                                       |

> 

Reviewer: Updated this table to stop mentioning OIDC-related stuff in 56a62782

## History

This proposal was originally known as 'OIDC-native' authentication.
To avoid the confusion with OIDC as a single sign-on/identity protocol, this proposal removed its dependency on OpenID Connect and was renamed to 'next-generation auth'.

Consistent with [MSC4186], this proposal can also be referred to as 'augmented atomic authentication', abbreviated to 'AAA'.

## Security considerations

Please refer to individual proposals.

## Unstable prefix

Please refer to individual proposals.

## Dependencies

> The following MSCs are part of this proposal:

Reviewer 2: from earlier: this phrasing makes it sound like the following MSCs are automatically accepted if this MSC is accepted, which feels unintentional. A slight wording change like the following helps clarify that other options are theoretically available, but the author (you) has put in the most work into these ones:

```suggestion
The following MSCs map the aforementioned OAuth 2.0/OIDC requirements to Matrix APIs:
```

- [MSC2964]
- [MSC2965]
- [MSC2966]
- [MSC2967]
- [MSC3824]
- [MSC4254]

> The following MSCs are not directly part of this proposal but this proposal assumes that these are accepted:

Reviewer 2: like above, slight wording change:

```suggestion
The following MSCs are not directly needed to achieve this proposal's overall objective, but are nice to have:
```

Reviewer: We did need those MSCs to happen for the implementation to work correctly. They were both merged in the meantime, so I can probably get rid of this section?

Reviewer: Did something in 152cd98f

[spec-auth]: https://spec.matrix.org/v1.13/client-server-api/#client-authentication
[spec-account]: https://spec.matrix.org/v1.13/client-server-api/#account-registration-and-management

As an example, on the Matrix.org homeserver, the user registration flow is gated behind three stages:

1. A CAPTCHA challenge
1. An agreement to the terms of service
1. An email verification

On paper, this lets clients implement their own UI for authentication, but in practice, the dynamic nature of the flows makes it difficult to implement a streamlined UI.
Moreover, the fact that those stages are displayed on client-driven UI, and not on a homeserver-specific domain, makes them insecure, impossible to design, or behave poorly in terms of user experience.

A few issues can be highlighted from this particular example:

- When setting up a CAPTCHA challenge, CAPTCHA services expect the challenge to be served from a specific domain. Because the client can be on any domain, [Synapse currently advises disabling host verification](https://element-hq.github.io/synapse/latest/CAPTCHA_SETUP.html#getting-api-keys).

>   When this option is disabled, the CAPTCHA service expects the server to verify the domain of the challenge, which is not possible.

Reviewer 5: Not sure if listing more is useful, but captcha design is specific to certain providers, support for 2FA would require modifications, etc.

Reviewer: I've tried to restrict the examples in the list only to this particular example (the matrix.org HS), which is why I haven't touched on those kind of details here

- The agreement to the terms of service can be completely ignored by the client, as it only requires the client to send a request with no specific data to validate it.
  This means the server has to trust the goodwill of the client to make sure the user agrees to the terms of service.
- The current design of the email verification flow has a confusing user experience.

>   The link sent to the user by email leads to a screen where the user is prompted to return to their client.

Reviewer 6: Isn't this a) how some MAS flows work, and b) is a relatively common paradigm?

Reviewer: The flow in MAS works by sending a code instead of a link for email verification. This has two advantages:

 - some platforms (like iOS/macOS) can autofill '2FA' codes got by email
 - this makes sure the user doesn't break the context of what they were doing. By making them enter a code, they will keep the browser window where they started the registration flow open while they check their email

Reviewer 7: > Things like security considerations, protecting client credentials, not re-inventing the wheel, 2FA, etc.

How are those first two achieved more when using OAuth, compared to the old way of doing things?

What if Google is forced by a government agency to keylog specific users' passwords entered into Matrix OAuth pages, as part of a secret operation against those users? Remember, Google Chrome is the most popular browser, and while parts of it are open source, the browser itself *is closed source*.

Relying on the security of a third party application means you have less ability to fix or even react to security holes that are found — or purposefully built into — those third party applications.

Besides, the arguably more important cryptographic keys that protect a user's messages and chat histories *have* to be known to the client *any*way, and *shouldn't* be known by the server. Matrix's whole design is centered around trusting the client more than the server, and requiring OAuth breaks that paradigm.

Reviewer: > What if Google is forced by a government agency to keylog specific users' passwords entered into Matrix OAuth pages, as part of a secret operation against those users? Remember, Google Chrome is the most popular browser, and while parts of it are open source, the browser itself is closed source.

I don't think this argument holds up well, as most Android users are probably using the default Google-provided keyboard, same with the iOS keyboard… and what prevents Windows from keylogging stuff on your screen anyway?

My point is, if we're at a point where we cannot trust the default browser on your computer, there are many other possible attack vectors to consider, with nothing we can realistically do to protect ourselves

Reviewer 8: It's true that it's a challenging threat model. But maybe lets try to not make things worse... Not every Matrix client is a web app, not every device to access Matrix is a smartphone etc. There are several scenarios where currently a Matrix client doesn't need to use (or trust) a huge 3rd-party application called a web browser. IMO it would be a big mistake to deprecate those scenarios. It's okay to use it for some cases (maybe even for majority cases), as in many (most?) cases it would be part of the Trusted Code Base anyway (as you noted above), but keep it possible to have an alternative.

> Besides, the arguably more important cryptographic keys that protect a user's messages and chat histories _have_ to be known to the client _any_way, and _shouldn't_ be known by the server. Matrix's whole design is centered around trusting the client more than the server, and requiring OAuth breaks that paradigm.

I don't think OAuth necessarily breaks this paradigm. But badly designed/implemented OAuth can.

Reviewer: The last messages in this thread don't relate to the initial concern; please open a new thread if you want to raise a concern that isn't covered by any of the currently opened thread.

---

Back to the original concern of this thread: I've added an example how the email flow could be better in 60e1d749 (which is how MAS works) and I think this section already explains what it tries to achieve: it takes one example in particular, highlighting what issues there are, noting that some of them *could* be improved with the current UIA paradigm

```md
A few issues can be highlighted from this particular example:

[…the bulletpoint list…–

**Note**: Many of these points could be improved with individual improvements to each of those stages, and multiple MSCs already exist to address some of them.
```

- Because the password field is displayed by the client, any system password manager will save the password as associated with the client domain.
  This means that if the user chooses to use a different client, they will have to manually find the password in their password manager, as the password will not be associated with the _service_, i.e., the homeserver domain.

**Note**: Many of these points could be improved with individual improvements to each of those stages, and multiple MSCs already exist to address some of them.

> ### Benefits of authenticating end-users through the system browser

Reviewer 6: What are the benefits of using the *system* browser, versus any old browser? Do we even need to be emphasising the use of system browser?

Reviewer: I'm using 'system browser' here as opposed to 'an embedded web view'. Maybe I should make it clearer by calling it the 'default browser'?

Reviewer 6: Ah right, then: why not an embedded web view?

Reviewer: Clarified that in 246ee1e1, closing this; reopen if it's not clear enough

Reviewer 6: Still, why the focus on "default" browser? Why not just say "a browser"?

Reviewer: You're right, any browser makes sense; did a5386b57

Rather than trying to fix the existing flows, this MSC proposes an alternative approach to authentication.
Authenticating end-users through the system browser is a well-established approach for many applications and would help solve most of the UI quirks mentioned above.
Though, some applications may wish to retain browser-less authentication, which this proposal supports thanks to the inherited authentication specifications.

> The general idea is simple: to authenticate a user, the client redirects the user to a URL on the homeserver, which completes the authentication flow, and then redirects the user back to the client.

Reviewer 5: Doesn't it really redirect to the IdP, not necessarily the homeserver?

Reviewer: In this iteration of the MSC, we don't separate the homeserver and the 'IdP'. Your client is logging in against your homeserver, with a different set of APIs than before; the fact that in practice it's a separate piece of software, maybe on a different domain, is an implementation detail?

This allows the homeserver to implement an authentication (and registration) flow tailored to its requirements.
A public service with open registration like Matrix.org could have secure CAPTCHA prompts, email verification, and terms of service agreements, with both password-based and social-based authentication.
A private service could have SSO-based authentication with specific 2FA requirements.

This has the benefit of working well with domain-bound authentication mechanisms:

- Password managers can be used to store the password for the homeserver domain, and the user can use a different client on a different domain without having to remember the password.
- WebAuthn credentials/Passkeys are bound to the domain, and would be impractical to introduce with the current authentication paradigm.

> - The user could benefit from sharing a single browser session with multiple clients, requiring them to enter their credentials only once per physical device.

Reviewer 6: Most of these feel like "nice to haves" rather than reasons to move to OAuth2, and as such I think actually hurt the argument. We wouldn't be making such a large change for these sorts of things. Honestly, I think we could either just get rid of these, or replace them with a paragraph at the end of the section, which brings the focus back on the the paragraphs that give the actual rationale for the move.

I think the same is true for the bullet points in the previous section.

Maybe something like:

> We also get other, more minor, benefits from the change to a more common auth mechanism. For example, password managers work out the box, it'd be possible to use webauthm/passkeys, etc.

Reviewer: That's fair. I think what I wanted to point out, is that using the homeserver's domain name as a base for all the auth flows has many small benefits. I've reworded a bit this section in 17244908, lemme know if that matches what you meant

Reviewer 6: I think we are still missing a high level, couple of sentences, proper rationale for moving to OAuth2 at the top. Currently the first concrete improvement mentioned is:

> When setting up a CAPTCHA challenge, CAPTCHA services expect the challenge to be served from a specific domain. Because the client can be on any domain, [Synapse currently advises disabling host verification](https://element-hq.github.io/synapse/latest/CAPTCHA_SETUP.html#getting-api-keys). When this option is disabled, the CAPTCHA service expects the server to verify the domain of the challenge, which is not possible.

which, I mean sure, is right, but feels *far* down the list of why we would want to do this.

To me, it seems like the overarching reason for OAuth2 is: to move to industry standard authentication protocol that supports full customisation by the homeserver.

Reviewer: I believe this was covered by 5bb8c579, please reopen if not

- Many other authentication mechanisms are effectively domain-bound: CAPTCHAs and client certificates, just to name a few.

This makes it possible to design widely different authentication flows for different homeservers, without having to cross an API boundary.
Implementers of said flows can focus on the specifics of their deployment without worrying about defining the right API between the client and the homeserver.

### Concealing the user's credentials

Another benefit of authenticating outside the client is that the client never has the user's full credentials.
This has two important implications:

- The client can't store the user's credentials, and thus can't use them to gain access without the user's consent.

> - The user can use different clients without worrying about revealing their account credentials to unknown parties. Only their homeserver ever interacts with their credentials.

Reviewer 5: Again -- isn't this really the IdP? Doesn't this also protect the homeserver ever knowing the user's password?

Reviewer: Same [as up there](https://github.com/matrix-org/matrix-spec-proposals/pull/3861/files#r1980966097); the fact that the homeserver and MAS are two pieces of software is an implementation detail

Reviewer 5: I think calling it the IdP doesn't imply it is a separate piece of software. 🤷

With the current state of the ecosystem, in a password-based account, this means sending the user's password again: nothing prevents the client from storing the password on the initial login and using it to perform these actions.
To put that in perspective, this means that if a user on the matrix.org homeserver tries to log in on a new client they want to try out, this client would be able to completely lock them out of their account by logging out other sessions and changing their password without additional confirmation from the user.

This also effectively widens the attack surface for credential theft, as both the client and the homeserver currently have access to the user's credentials.

Making it mandatory for the client to go through the system browser to authenticate means there is a part of the login flow that the client can't skip and doesn't have control over.
The user has to give their explicit consent during that part of the flow, with no way for the client to bypass it.

This opens up the possibility of giving more restrictive access to the user's account.

> It would open up Matrix to a new class of clients with only partial access to the user's account.

Reviewer 5: It feels like this should be prominent in the proposal.

Reviewer: I agree! I've added a sentence about it in add4799d. wdyt?

This MSC intentionally does not cover the use of alternative authorization flows that don't rely on a web browser, for automation or other purposes.
This does not mean that the Matrix ecosystem should not embrace such flows, but it is rather an attempt to keep this proposal focused on making this a good default flow for most users, ensuring it is treated as a first-class citizen in the ecosystem.

The goal is to set a new widely-adopted base for authentication in the Matrix ecosystem, eventually replacing the current custom authentication protocol.
Solving Matrix-specific problems with this new base could benefit the wider ecosystem of decentralized protocols, rather than staying confined to Matrix.

### Why not 'just use OpenID Connect'?

OpenID Connect does a good job at standardizing on top of OAuth 2.0, and it covers most things happening between the client and the server for authentication.
It is a great fit for connecting identity providers to other pieces of software, and this is already what homeservers do with the [`m.login.sso`] flow.

Knowing that, it can feel like adopting OpenID Connect fully would help using off-the-shelf identity providers for Matrix homeservers.

> In practice, OpenID Connect does not cover continuous exchanges between the application and the identity providers: there is no well-supported standard to signal new sessions, new users, sessions ending, users deactivation, etc. from the identity provider to the application.

Reviewer 5: I can make some assumption, but why do we care about this in Matrix?

Reviewer: Explained that better in 559762de, please reopen if it's not clear enough

Reviewer 5: Would it be fair to say that the impact for Matrix is mostly only servers and clients will be able to mostly implement "normal" OAuth2?

Reviewer: That is correct; this is why the sentence right before talks about 'off-the-shelf identity providers for Matrix *homeservers*'?

Reviewer 5: I think that's a bit subtle, but fair enough!

**Note**: an earlier version of this MSC focused on 'delegating' authentication to an identity provider, but it showed its limitations and added much confusion over the intent of the proposal.

### Keeping the ecosystem open

One common critique of OAuth 2.0 and OpenID Connect is that they are widely used in contexts where the service provider controls which clients are allowed to interact with the service.
This usually implies a contractual relationship between the service provider and the client, typically through a developer program, where the client must comply with the service provider's terms of service.

This has been a notorious problem with [OAuth 2.0 in email protocols][thunderbird-oauth2], where email clients are forced to register their applications with each email provider, giving the email provider the right to reject any application.

The goal of this proposal is to keep the ecosystem open by clearly defining how clients register themselves dynamically with the homeserver.

> This won't prevent the homeserver from rejecting clients arbitrarily, but this was already the case with UIA.

Reviewer: Explained that in 6f74515f; reopen if this is not clear enough

> On the contrary, this proposal will provide homeserver operators greater visibility into which clients are being used, helping with decisions to approve third-party clients.

Reviewer: What I was trying to say here is "if you restrict third party clients (which you could definitely already do before), now you have better visibility of what clients people are trying to use, so you can make decisions to open up more your deployment". Does that make sense?

> On the contrary, this proposal will provide homeserver operators greater visibility into which clients are being used, helping with decisions to approve third-party clients.

Reviewer 4: If our concern is that admins block out clients they don't like, how does giving them more visibility about which clients people are using help?

I've commented at https://github.com/matrix-org/matrix-spec-proposals/pull/3861#discussion_r1989579374. Resolving this thread in favour of that one.

[thunderbird-oauth2]: https://wiki.mozilla.org/Thunderbird:Autoconfiguration:ConfigFileFormat#OAuth2

## Proposal

This proposal introduces a new set of authentication APIs for Matrix, based on OAuth 2.0 and OpenID Connect (OIDC) specifications.

As a first step, it introduces those APIs as alternatives to the existing User-Interactive Authentication (UIA) APIs, acknowledging the complexity of covering all the use cases of the existing APIs.

> The long-term goal is to deprecate the existing UIA APIs and replace them with the new OAuth 2.0/OIDC-based APIs.

Reviewer 4: > as alternatives to the existing User-Interactive Authentication (UIA) APIs 
>  deprecate the existing UIA APIs

and `/login`, right?

(and `/logout`, maybe?)

Reviewer 4: To be honest, the term "UIA APIs" isn't very clear... there are APIs which currently *use* UIA, but I'm not sure if you plan to remove them all, or just stop them using UIA. Would be good to be more explicit here.

Reviewer 5: And `/register`?

Reviewer: Clarified what I meant in e797837a. Is that better?

Reviewer 4: > And `/register`?

`/register` *is* a UIA API, so is covered by the existing wording.

Reviewer 4: > Clarified what I meant in [e797837](https://github.com/matrix-org/matrix-spec-proposals/commit/e797837a64bf3793e2fc2f109441944718372898). Is that better?

Not really, I'm afraid. I've commented at https://github.com/matrix-org/matrix-spec-proposals/pull/3861#discussion_r1989579374. Will close this thread.

#### Application services

In the longer term, application services could leverage alternative grant types like the [OAuth 2.0 Client Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.4) to obtain access to the homeserver.
In the meantime, homeservers should keep registration through the [`/_matrix/client/v3/register` with the `m.login.application_service` type][register-app-service] endpoint working for application services.
[MSC4190: Device management for application services][MSC4190] proposes a simple API to create and delete devices for users managed by application services, to remove the need for keeping the [`m.login.application_service`] login type working.

[register-app-service]: https://spec.matrix.org/v1.13/application-service-api/#server-admin-style-permissions

## Sample flow

This section describes a sample flow, taking together the steps described in more details in the dedicated MSCs.
It assumes the client already discovered the homeserver's Client-Server API endpoint.

[areweoidcyet.com](https://areweoidcyet.com/client-implementation-guide/) has an interactive guide on how to use this flow.

### Discovery [MSC2965]

> First step is to discover the homeserver's authorization server metadata.

Reviewer 9: Struggling a little with the duplication between MSCs here but I see Erik has already asked what I was going to ask at https://github.com/matrix-org/matrix-spec-proposals/pull/2965/files#r1931976488

Reviewer: This part is there to have a full authentication flow described *somewhere*. This is whole MSC does not specify anything, but rather explains what the direction is, and how the sub-MSCs fit together to have a full 'next-gen' authentication flow, if that makes sense?

Reviewer 9: OK, I'm not sure about the idea of an MSC that doesn't specify anything (feels like it could just be an ancillary doc of some kind) but it seems okay.

[MSC3262]: https://github.com/matrix-org/matrix-spec-proposals/pull/3262
[MSC3741]: https://github.com/matrix-org/matrix-spec-proposals/pull/3741
[MSC3744]: https://github.com/matrix-org/matrix-spec-proposals/pull/3744
[MSC3782]: https://github.com/matrix-org/matrix-spec-proposals/pull/3782
[MSC3824]: https://github.com/matrix-org/matrix-spec-proposals/pull/3824
[MSC3967]: https://github.com/matrix-org/matrix-spec-proposals/pull/3967
[MSC3970]: https://github.com/matrix-org/matrix-spec-proposals/pull/3970
[MSC4186]: https://github.com/matrix-org/matrix-spec-proposals/pull/4186
[MSC4190]: https://github.com/matrix-org/matrix-spec-proposals/pull/4190
[MSC4191]: https://github.com/matrix-org/matrix-spec-proposals/pull/4191
[MSC4198]: https://github.com/matrix-org/matrix-spec-proposals/pull/4198
[MSC4254]: https://github.com/matrix-org/matrix-spec-proposals/pull/4254
[`m.login.application_service`]: https://spec.matrix.org/v1.13/client-server-api/#appservice-login
[`m.login.sso`]: https://spec.matrix.org/v1.13/client-server-api/#single-sign-on
[`/_matrix/client/v3/capabilities`]: https://spec.matrix.org/v1.13/client-server-api/#get_matrixclientv3capabilities

> [`/_matrix/client/v3/login`]: https://spec.matrix.org/v1.13/client-server-api/#post_matrixclientv3login

Reviewer 9: Having an overview like this is incredibly useful and thanks so much for putting it together, although I am struggling to pick out what this MSC is actually intending to change about the spec between the duplicated bits of other MSCs. Is it largely the glue of how they various different MSCs fit together?

Reviewer: This isn't supposed to be an overview, rather linkifying a bunch of MSCs used in the text. This lets me use this markdown syntax to link: `[MSC3824: OIDC-aware clients][MSC3824]`. The overview is in the `Proposal` section of the text

Reviewer 10: @sandhose I think Dave was referring to this MSC as a whole, rather than just this portion of the document (n.b. @dbkr I'd suggest using the "comment on this file" button for this types of questions in future to avoid confusion:)

![image](https://github.com/user-attachments/assets/b731c32b-14ed-481a-b352-2ff97d2dd3b9)

@sandhose simply put, I believe the question is "if I were writing a spec PR for this MSC... what would I write?".

WebAuthn credentials/Passkeys, as well as client certificate-based authentication, are all bound to the domain they are registered on, making them impractical if the client is directly authenticating the end user.
Password managers will also function better if the credentials are bound to the homeserver domain instead of the client itself.

This makes it possible to design widely different authentication flows for different homeservers, without having to cross an API boundary.
Implementers of said flows can focus on the specifics of their deployment without worrying about defining the right API between the client and the homeserver.

Bouncing between the client and the browser may lead to user confusion, especially on operating systems with limited window management capabilities.
Mobile operating systems such as iOS and Android provide a way to embed a secure browser view within an application ([`ASWebAuthenticationSession`](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession)) on iOS, [Custom Tabs](https://developer.android.com/develop/ui/views/layout/webapps/overview-of-android-custom-tabs) on Android).

> In those cases, the host application cannot control or monitor what is happening within the embedded browser view, protecting the user's credentials, but that view shares the same context as the system-wide browser.

Reviewer: I think it's still important to point out what those API provides, why they are safe to use compared to web views?

> In those cases, the host application cannot control or monitor what is happening within the embedded browser view, protecting the user's credentials, but that view shares the same context as the system-wide browser.

Reviewer: Prehaps we can just drop the 'protecting the user's credentials' part?

```suggestion
In those cases the host application cannot control or monitor what is happening within the embedded browser view, but that view shares the same context as the system-wide browser.
```

### Concealing the user's credentials

Another benefit of authenticating outside the client is that the client never has the user's full credentials.

> This has two important implications:

Reviewer 6: I'm wondering if this is actually true for a nefarious client? If they want they can embed a browser that the app can read the content of? I don't think anything here stops that?

Reviewer 11: in the end, if your client is nefarious, it's game over: with or without OIDC, it could steal your credentials or your conversation history or otherwise phish you. So, perhaps the right fix here is to clarify that non-malicious clients will never get the user's full credentials (but obviously OIDC doesn't protect against malicious clients).

Reviewer 6: The text is asserting stronger guarantees than you can generally get, and then trying to use that to justify the move to OAuth2. Certainly its nice that in certain cases we can make it harder for bad clients to steal credentials, but we can't really claim too much here.

I think the reason that this (and the other bits) grate for me when reading is that all of these sections are under "Motivation", which starts by saying:

> To understand this proposal, it is important to lay out the reasoning behind it. ... This section explains this reasoning before getting into the detailed proposal.

I think its great that we are documenting why we are doing this, but I don't think these sections are really *why* we want to move to OAuth2, but are instead listing some nice benefits of doing so.

If we are going to give a rationale, then for documentation purposes I think its important that its accurate. Equally, I don't want to bog down this MSC by suggesting lots of changes that don't change the practical proposal.

As such, I propose we add a "Rationale" section before the current "Motivation" section that says something along the lines of (but with better English):

> The goal of this MSC is to propose a new set of authentication APIs for Matrix, based on the OAuth 2.0 and OpenID Connect (OIDC) specifications. To understand this proposal, it is important to lay out the reasoning behind it.
>
> ### Rationale
>
> Matrix currently uses a custom protocol for authentication, one that hasn't been updated in many years. This leads to two main problems:
>   1.  The custom protocol does not follow best practices, nor does it use the collective experience of other authentication protocols.
>   2. While relatively generic, the custom protocol does not currently support many desired features that users and organisations actually want.
>
> These points could be addressed by iterating on the current protocol, however using the industry standard OAuth2 allows us to benefit from their experience without having to build a competing (relatively generic) authentication protocol. We also will benefit more easily from any further enhancements, or best practice recommendations, of OAuth2.
>
> ### Background and motivation
> 
> This section expands on the benefits of moving to OAuth2. It is purely informative and not part of the actual proposal. The proposal itself is high-level, and the technical implications are split into separate, more focused MSCs.
>
> ...

Which I think a) gives a (hopefully) accurate enough summary of the core rationale for moving to OAuth2, and b) allows us to keep the current sections that allow some of the problems of the existing protocol and benefits to moving to OAuth2.

Reviewer 6: (Otherwise I'm happy with the meat of the proposal)

Reviewer: Thanks a lot for the suggestion. I've incorporated it in 5bb8c579

This can be mitigated to some extent by having homeservers implement authorization flows that don't require JavaScript, but this doesn't completely eliminate the risk.

### Homeserver implementation complexity

In practice, to provide a good user experience, homeservers have to implement web views for the authentication flows, which is complex to implement well.
This means having proper accessibility, translations, and UX.
Those concerns were previously mainly affecting client implementations, and will now also affect homeserver implementations.

On the other hand, the previous registration flow was notoriously complex to implement both for clients and homeservers, and this proposal removes a lot of that complexity from the client side.

Moreover, these concerns should, in theory, already apply to the homeserver-side implementation, as the homeserver is supposed to provide a web-based fallback for all UIA steps.
In practice, this is often not the case, with either missing fallbacks for some UIA steps or sub-par user experiences in their implementations.

> ## Alternatives

Reviewer 11: I was a bit surprised that we don't explicitly call out (independently of the reference to MSC3262 and friends below) that MSC3861 explicitly does not try to support using the same password for both login and encryption recovery.

Specifically, it'd be useful to spell out that when OIDC is hooked up to an existing SSO mechanism, there would be no way to link encryption recovery to your SSO credentials - and so separate recovery passphrase would always be needed for those instances.  In other words, this is a trade-off between supporting SSO or other flexible authentication... and having fancy PAKE style auth, and so for now we are consciously trading off in favour of SSO (especially given the need to support enterprise SSO auth envs).

Given SSO is incompatible with PAKE, you'd always have to treat PAKE-style approaches as a distinct auth mechanism anyway, which is why it's acceptable to punt it to a separate custom grant (and so need a new MSC to define a special PAKE-compatible Resource Owner Password Credentials flow).

Reviewer 11: (n.b. this isn't remotely a blocking concern; i just think it would be useful to capture the full rationale for the design in the overview MSC if poss)

| [MSC3782: Matrix public key login spec][MSC3782]                                                                                | Similar to above                                                                                                                                                                                                                                                                                                                 |
| [MSC3744: Support for flexible authentication][MSC3744]                                                                         | This proposal would instead be used as the pluggable layer for auth in Matrix                                                                                                                                                                                                                                                    |

## History

This proposal was originally known as 'OIDC-native' authentication.
To avoid the confusion with OIDC as a single sign-on/identity protocol, this proposal removed its dependency on OpenID Connect and was renamed to 'next-generation auth'.

> Consistent with [MSC4186], this proposal can also be referred to as 'augmented atomic authentication', abbreviated to 'AAA'.

Reviewer 3: Please no, it overlaps with ["Authentication, authorization, and accounting" paradigm](https://en.wikipedia.org/wiki/Authentication,_authorization,_and_accounting), security people widely use the same acronym for it.

Reviewer: That was a joke/easter egg, as MSC4186 has a notoriously bad acronym, SSS for 'simplified sliding sync', but has nothing 'sliding' about

Reviewer: Removed in 1706907112fc3fd98b09e79dba91cbc5487451a3

Reviewer 3: Oh, now I understand. Sorry - my sense of humour is known to be flawed ;(

## History

This proposal was originally known as 'OIDC-native' authentication.
To avoid the confusion with OIDC as a single sign-on/identity protocol, this proposal removed its dependency on OpenID Connect and was renamed to 'next-generation auth'.

## Security considerations

Please refer to individual proposals.

## Unstable prefix

Please refer to individual proposals.

## Dependencies

The following MSCs map the aforementioned OAuth 2.0/OIDC requirements to Matrix APIs:

- [MSC2964: Usage of OAuth 2.0 authorization code grant and refresh token grant][MSC2964]
- [MSC2965: OAuth 2.0 Authorization Server Metadata discovery][MSC2965]
- [MSC2966: Usage of OAuth 2.0 Dynamic Client Registration][MSC2966]
- [MSC2967: API scopes][MSC2967]
- [MSC4254: Usage of RFC7009 Token Revocation for Matrix client logout][MSC4254]

> The following MSCs are meant to cover less escential parts of the authentication and account management flows.

Reviewer 10: ```suggestion
The following MSCs are meant to cover less essential parts of the authentication and account management flows.
```

**Note**: an earlier version of this MSC focused on 'delegating' authentication to an identity provider, but it showed its limitations and added much confusion over the intent of the proposal.

### Keeping the ecosystem open

One common critique of OAuth 2.0 and OpenID Connect is that they are widely used in contexts where the service provider controls which clients are allowed to interact with the service.
This usually implies a contractual relationship between the service provider and the client, typically through a developer program, where the client must comply with the service provider's terms of service.

This has been a notorious problem with [OAuth 2.0 in email protocols][thunderbird-oauth2], where email clients are forced to register their applications with each email provider, giving the email provider the right to reject any application.

The goal of this proposal is to keep the ecosystem open by clearly defining how clients register themselves dynamically with the homeserver.
This won't prevent the homeserver from rejecting clients arbitrarily, but this was already the case with UIA.

> On the contrary, this proposal will provide homeserver operators greater visibility into which clients are being used, helping with decisions to approve third-party clients.

Reviewer 2: non-blocking: and helps homeserver operators make decisions about their homeservers, like when to roll out features which break in some clients.

This has been a notorious problem with [OAuth 2.0 in email protocols][thunderbird-oauth2], where email clients are forced to register their applications with each email provider, giving the email provider the right to reject any application.

The goal of this proposal is to keep the ecosystem open by clearly defining how clients dynamically register themselves with the homeserver.

> While this won't prevent a homeserver from arbitrarily rejecting clients (which is already possible with UIA based on the client's user-agent string or redirect URI, for example), this proposal will provide homeserver operators greater visibility into which clients are being used, aiding decisions to approve third-party clients.

Reviewer: > which is already possible with UIA

I meant 'legacy' /login in this case, fixed that in 9df174c7

> While this won't prevent a homeserver from arbitrarily rejecting clients (which is already possible with UIA based on the client's user-agent string or redirect URI, for example), this proposal will provide homeserver operators greater visibility into which clients are being used, aiding decisions to approve third-party clients.

Reviewer: > Sorry, but I still don't think your updated text is clear here. It still seems to be in conflict with the previous paragraph where it says that it is problematic to let server admins decide which clients to allow.

How about 8e114016?

This should be read as:

 - usually people don't like OAuth 2.0/OIDC in open protocols because it requires a contract between the client and the service (e.g. a manual registration in a dev portal)
 - to mitigate this, we define a way to dynamically register the client

> While this won't prevent a homeserver from arbitrarily rejecting clients (which is already possible with UIA based on the client's user-agent string or redirect URI, for example), this proposal will provide homeserver operators greater visibility into which clients are being used, aiding decisions to approve third-party clients.

Reviewer: After discussing this internally, I simplified this in a379a33f

[thunderbird-oauth2]: https://wiki.mozilla.org/Thunderbird:Autoconfiguration:ConfigFileFormat#OAuth2

## Proposal

This proposal introduces a new set of authentication APIs for Matrix, based on OAuth 2.0 and OpenID Connect (OIDC) specifications.

> As a first step, it introduces those APIs as alternatives to the existing User-Interactive Authentication (UIA) APIs, both the ones used in authentication and UIA-protected APIs related to account management, acknowledging the complexity of covering all the use cases of the existing APIs.

Reviewer: I keep using 'UIA' as '/login', sorry for the confusion 😅
